jerry-ext/jerry-context/include/jerry-context.h

```diff
@@ -0,0 +1,43 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRYX_CONTEXT_H
+#define JERRYX_CONTEXT_H
+
+#include "jerryscript.h"
+
+#ifndef JERRYX_CONTEXT_SLOTS
+#define JERRYX_CONTEXT_SLOTS 1
+#endif /* !JERRYX_CONTEXT_SLOTS */
+
+/**
+ * Function to pass to jerry_init_with_user_context() as its second parameter
+ */
+void *jerryx_context_init (void);
+
+/**
+ * Function to pass to jerry_init_with_user_context() as its third parameter
+ */
+void jerryx_context_deinit (void *user_context_p);
+
+/**
+ * Function to request an index for per-context data
+ */
+int jerryx_context_request_slot (jerry_user_context_init_cb init_cb,
+                                 jerry_user_context_deinit_cb deinit_cb);
```

Comment:
As an alternative to requesting/getting "slots" at runtime, I think it's also possible to do this at compile/build time *without having a single file that defines/lists the slot IDs*:

By defining a dedicated linker section, i.e. `.jerryx_context_slots`, into which to put each `static jerryx_context_slot_info_t`. The linker script can add a terminator at the end of the section with `NULL` values. The linker script should also emit a symbols with the base address and also the size of the section, so that the "slot" indices and also the total number of slots can be calculated at build time. Probably `jerry-ext/...` should include a linker script snippet defining the section + symbols, so it can be `INCLUDE`d into the .ld script by the user of the util.

Doing this vs runtime requesting/getting would save code space and probably be a bit faster as well.

Now, admittedly, this solution is probably a bit more complicated to use, because it involves having to add a couple things to the target's linker script.

So, I think having a simple implementation (your one) has value.

Ideally, I would like to be able to swap out your implementation with the linker script based one **without having to change the code that uses the util**. But I think the current API doesn't let me do this easily.

Runtime registration implementation:
```
static int my_slot = -1;
// register at boot:
{
  my_slot = jerryx_context_request_slot (my_init, my_deinit);
}
// get context:
{
  void *ctx = jerryx_context_get_slot (my_slot);
}
```

Compile-time registration implementation:
```
static const jerryx_context_slot_info_t my_context_cbs __attribute__((section(".jerryx_context_slots"))) = {
  .init = my_init,
  .deinit = my_deinit,
};

extern jerryx_context_slot_info_t *jerryx_context_slots_first; // defined in linker script, todo: move to jerry-context.h
// note const! -- when const, the slot value in jerryx_context_get_slot() can be an immediate vs having to load it.
static const int my_slot =  &my_context_cbs - jerryx_context_slots_first; // todo: put macro in jerry-context.h

// get context -- still the same:
{
  void *ctx = jerryx_context_get_slot (my_slot);
}

// the implementation of jerryx_context_get_slot() can be simpler,
// no checking of the slot arg is needed:
inline void *jerryx_context_get_slot(int slot) {
  return ((void **) jerry_get_user_context ())[slot];
}
```

Thoughts? Ideas on how we could have one interface that enables both implementations?
@zherczeg @akosthekiss Thoughts on the linker script trickery I'm suggesting here?